### ISSUE TITLE
Make an optional variable for easy loading of unstructured grids in vtk-compatible format

### ISSUE THREAD

**Issue by Boorhin** [ugrid-conventions/ugrid-conventions#56]
VTK is quite a widespread format and I use it a lot, in particular with pyvista.
[Pyvista needs 2 information](https://docs.pyvista.org/examples/00-load/create-unstructured-surface.html#unstructuredgrid-with-shared-points), the node coordinate table (vertice) which is basically in UGrid `node_x`, `node_y` and `node_z` as triplets
And a connectivity table (cells) which is close to UGRID `volume_node_connectivity`.
However, the vtk format needs an extra information even for unstructured grid which is the number of nodes in one cell for each cell for example of triangular mesh the array would be
vtk_cell= [[3,1,2,3],
[3,12,1,5],
...,
[3, 3,24,4],
[3, 23,24,6]]
you can simply generate it with:

``` python
import pyvista
node_z= compute_z_of_each_node_return_a_1D_array()
vertices=np.vstack((node_x, node_y, node_z)).T
vtk_cell= np.hstack((np.ones[:,None]*6,volume_node_connectivity))
pa=pyvista.UnstructuredGrid(cells=vtk_cell, points= vertices)
```

For more complicated geometries with various cell-types, it is better to use `pyvista.PolyData` that will work in a similar way.
I thus recommend to have volume_node_connectivity as an optional argument in UGRID for unstructured meshes.
of course one could argue that you can compute the cell argument of vtk with `face_nodes` and the dimension 'layer' but it is quite tedious.

**Comment by hrajagers** [comment]
1. Since it's so easy to generate from the `face_node_connectivity` array, why put it into the file?
2. Adding an array to store the number of nodes just duplicates information already in the file and duplicating information leads to potential inconsistency. In the beginning of UGRID, we have discussed whether to use the an numFaces x maxNumNodes array with missing values for unused node entries, or a sparse storage with one array for the number of nodes per face and another linear array with all the node indices for all the faces. For UGRID we selected the former method as it seemed to be more consistent with the current netCDF usage and the variation in the number of nodes per cell is typically limited. For the [CF Geometries](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.9/cf-conventions.html#geometries) the developers have switched to the latter approach since polylines and polygons can easily vary by an order of magnitude causing a lot of overhead in the former method.
3. UGRID does already have a `volume_node_connectivity` attribute, but it's ony used for the [fully 3D unstructured topologies](http://ugrid-conventions.github.io/ugrid-conventions/#fully-3d-unstructured-ie-non-layered-mesh-topology). For those topologies we have used concepts consistent with other 3D storage formats and hence also more naturally compatible with vtk. The number of nodes per volume is not explicitly specified, but a direct consequence of the selected volume type. Note that for consistency the `volume_node_connectivity` is still a numVolumes x maxNumNodes array and not a linear array. The traditional Telemac Selafin/Serafin format tends a bit more towards that format. All 3D Telemac volumes are of the type `wedge`: a prism with triangular base. (I just noticed that the UGRID table of 3D elements is not formatted correctly; I will try to correct that)
4. Adding an optional `volume_node_connectivity` attribute to the 2D mesh would require a linear index for all 3D nodes. Currently we don't have such a dimension and hence we would first have to agree on how those nodes should be numbered: column first or plane first, bottom first or top first. Probably different users would prefer different numbering order and hence we would need a way to encode that, and all the post-processing tools would have to process that correctly.

**Comment by Boorhin** [comment]
`face_node_connectivity` is not easy to generate, it needs to be created when you generate your mesh and it is already in UGrid for unstructured grids. I suppose you refer to `volume_node_connectivity`. I don't think it is so easy to generate but let say it is not too complicated but resource intensive
Numbering convention for nodes is bottom-up in CF if I remember correctly. Actually indexing the nodes in 3D is very straightforward. It is a simple matrix addition. I have not found a vectorial way of making a` volume-node-connectivity` without going through a loop (doesn't mean it is not possible). I just thought that since the information was available in some formats why not put it in as optional to help loading the meshes.

**Comment by hrajagers** [comment]
I'm not sure that I follow your arguments.

I didn't say that the `face_node_connectivity` is easy to generate, I only indicated that the extra information required by vtk (the number of nodes in one cell) is easy to generate from the available data. I thought that in your original mail you suggested to include the extra information, but I would suggest not to do so for the reasons 1 and 2 in my previous post. I guess we agree that `face_node_connectivity` is already available in the UGRID file.

I don't think that CF requires bottom-up numbering of nodes, but that's a minor point. One could add an enumeration of different options.

If we use a separate dimension for the vertical then the indexing of the volumes is trivial. For each volume at level _k_ in column _i_ the `face_node_connectivity(i,:)` gives the first index of the corner nodes, and _k_ and _k+1_ give the second index of the corner nodes. However, you seem to prefer a linear index over two indices. Because that's more consistent with vtk ... Correct?

So, the `volume_node_connectivity` would require a linear indexing of the 3D nodes. If I interpret your post properly, you propose to use say (1-based) indices 1 to nNodes2D for the bottommost layer, nNodes2D+1 to 2*nNodes2D for the first layer up, etc. up to the final laye (nLayers-1)*nNodes2D+1 to nLayers*nNodes2D. Correct?
Would you want to store the data at the nodes in a two-dimensional (nNodes, nLayers) array (ignoring optional time dimension) or in a one-dimensional array of length nNodes*nLayers?

Whether the `volume_node_connectivity` can be created from the existing `face_node_connectivity` without loops depends on the environment in which you try to do the conversion. If `FaceNodeConn` is an nFaces x maxNFaceNodes array in MATLAB with NaN values for dummy node entries, e.g. the 4th node of a triangle, and nLayers is the number of node layers (nLayers-1 volume layers), then something like the following code would do the trick:
```
nNodes2D = max(FaceNodeConn(:); % count the number of nodes
nFaceNodes = sum(~isnan(FaceNodeConn),2); % count the number of nodes per face
FaceNodeConn_t= FaceNodeConn'; % transpose the connectivity array such that each column represents a face

VolNodeConn1 = [FaceNodeConn_t; FaceNodeConn_t+nNodes2D]; % add a second set of nodes for the volumes of layer 1
LinVNC1 = VolNodeConn1(~isnan(VolNodeConn1)); % create a linear array of node numbers for layer 1
nVolNodes1 = 2*nFaceNodes; % number of nodes per volume (for one layer)

AllVNC = LinVNC1 * ones(1,nLayers-1) + nNodes2D * ones(2*nNodes2D,1) * (0:nLayers-2); % generate indices for all layers
LinVNC = AllVNC(:); % convert to a linear index of the nodes for all volumes
nVolNodes = repmat(nLayers-1,1,nVolNodes1); % number of nodes per volume (all layers)
```
Something similar is likely possible using Python/NumPy, but may not be possible in vtk. Either way, I expect that adding the `volume_node_connectivity` isn't the only adjustment needed and hence it would be good to create a small example file or an example ncdump for such a mesh with data. It may either be a valuable addition to the 2D/3D UGRID conventions, or it might be a best practices recommendation for model developers that want to enable users to visualize the results using vtk based tools.

**Comment by ChrisBarker-NOAA** [comment]
>it would be good to create a small example file or an example ncdump for such a mesh with data.

yes, please, I know I'm still a bit confused as to exactly what is being asked for here. Concrete examples help a lot :-)

> It may either be a valuable addition to the 2D/3D UGRID conventions, or it might be a best practices recommendation for model developers that want to enable users to visualize the results using vtk based tools.

Agreed -- after all, there is nothing in either the UGRID or CF standard that prohibits additional data from being included.
